src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java

```diff
@@ -0,0 +1,1753 @@
+      DatabaseMetaData metadata,
+      Set<String> types
+  ) throws SQLException {
+    // Compute the uppercase form of the desired types ...
+    Set<String> uppercaseTypes = new HashSet<>();
+    for (String type : types) {
+      if (type != null) {
+        uppercaseTypes.add(type.toUpperCase());
+      }
+    }
+    // Now find out the available table types ...
+    Set<String> matchingTableTypes = new HashSet<>();
+    try (ResultSet rs = metadata.getTableTypes()) {
+      while (rs.next()) {
+        String tableType = rs.getString(1);
+        if (tableType != null && uppercaseTypes.contains(tableType.toUpperCase())) {
+          matchingTableTypes.add(tableType);
+        }
+      }
+    }
+    return matchingTableTypes.toArray(new String[matchingTableTypes.size()]);
+  }
+
+  @Override
+  public IdentifierRules identifierRules() {
+    if (identifierRules.get() == null) {
+      if (defaultIdentifierRules != null && connections.peek() == null) {
+        // We've not yet established a connection to the database, so in this case
+        // just return the default rules ...
+        return defaultIdentifierRules;
+      }
+      // Otherwise try to get the actual quote string and separator from the database, since
+      // many databases allow them to be changed
+      try (Connection connection = getConnection()) {
+        DatabaseMetaData metaData = connection.getMetaData();
+        String leadingQuoteStr = metaData.getIdentifierQuoteString();
+        String trailingQuoteStr = leadingQuoteStr; // JDBC does not distinguish
+        String separator = metaData.getCatalogSeparator();
+        if (leadingQuoteStr == null || leadingQuoteStr.isEmpty()) {
+          leadingQuoteStr = defaultIdentifierRules.leadingQuoteString();
+          trailingQuoteStr = defaultIdentifierRules.trailingQuoteString();
+        }
+        if (separator == null || separator.isEmpty()) {
+          separator = defaultIdentifierRules.identifierDelimiter();
+        }
+        identifierRules.set(new IdentifierRules(separator, leadingQuoteStr, trailingQuoteStr));
+      } catch (SQLException e) {
+        if (defaultIdentifierRules != null) {
+          identifierRules.set(defaultIdentifierRules);
+          log.warn("Unable to get identifier metadata; using default rules", e);
+        } else {
+          throw new ConnectException("Unable to get identifier metadata", e);
+        }
+      }
+    }
+    return identifierRules.get();
+  }
+
+  @Override
+  public ExpressionBuilder expressionBuilder() {
+    return identifierRules().expressionBuilder();
+  }
+
+  /**
+   * Return current time at the database
+   *
+   * @param conn database connection
+   * @param cal  calendar
+   * @return the current time at the database
+   */
+  @Override
+  public Timestamp currentTimeOnDB(
+      Connection conn,
+      Calendar cal
+  ) throws SQLException, ConnectException {
+    String query = currentTimestampDatabaseQuery();
+    assert query != null;
```

Comment:
nit: Is `assert query != null && !query.isEmpty();` too condensed? I'm used to see those two checks together.

Reply:
If the assertion fails when they are combined, you don't know whether it is null or empty. By having separate assertions, when it does fail you know exactly why.